### DISCUSSION TITLE
How do you handle the scoping of parsers?

### DISCUSSION THREAD

**Discussion by sillygoose** [pointfreeco/swift-parsing#283] [Q&A] [answered]
Just wondering how others handle the scoping of parser names.  I only have one parser for the time being but as I add new devices I will have to add new parsers to handle their output.  Right now the high-level parsers are global scope and for organization split across multiple source files, the lower level parsers used by the top-level parser are all private so not a problem.

I can't use a struct to organize them as computed properties are not allowed in extensions so just wondering what the experts do in production code to control the scope of parser names and keep them isolated.

**Comment by sillygoose** [comment, marked as answer]
I tried some ideas and I was able to use simple structures for organizing the parser code, once I found that static computed properties are permissible in extensions.  So I get to control my variable scopes and organize the parser source files.

**Reply by mbrandonw** [reply]
Hi @sillygoose, glad you got something working that you are happy with. We also plan on revisiting parsers sometime soon to bring better ergonomics to them, such as using the `body` style of parsing:

```swift
struct UserParser: Parser {
  var body: some Parser<Substring, User> {
    Parse(User.init) {
      Int.parser()
      ","
      Prefix { $0 != "," }
      ","
      Bool.parser()
    }
  }
}
```

This style is more performant, plays nicely with recursive parsers, and type checks a lot faster since Swift doesn't have to deal with gigantic generic types.

**Reply by sillygoose** [reply]
I'll look forward to your update, I made the decision to go with Point-Free to jump start my Swift/SwiftUI learning adventure and it just keeps paying off.
